client/Android/Studio/freeRDPCore/src/main/java/com/freerdp/freerdpcore/services/LibFreeRDP.java

```diff
@@ -273,6 +275,61 @@ public static boolean setConnectionInfo(int inst, BookmarkBase bookmark) {
         String[] arrayArgs = args.toArray(new String[args.size()]);
         return freerdp_parse_arguments(inst, arrayArgs);
     }
+    
+    public static boolean setConnectionInfo(int inst, Uri openUri) {
+        ArrayList<String> args = new ArrayList<String>();
+        HashMap<String, String> map = new HashMap<String, String>();
+
+        // Parse URI from query string. Same key overwrite previous one
+        // freerdp://user@ip:port/title?sound=&rfx=&p=password&clipboard=%2b&themes=-
+     
+        // Now we only support Software GDI
+        map.put("gdi", "/gdi:sw");
+        
+        // Parse hostname and port. Set to 'v' argument
+        String hostname = openUri.getHost();
+        int port = openUri.getPort();
+        if (hostname != null) {
+            hostname = hostname + ((port == -1) ? "" : (":" + String.valueOf(port)));
+            map.put("v", "/v:" + hostname);
+        }
+        
+        String user = openUri.getUserInfo();
+        if (user != null) {
+            map.put("u", "/u:" + user);
+        }
+        
+        for (String key: openUri.getQueryParameterNames()) {
+            String value = openUri.getQueryParameter(key);
+            
+            if (value.isEmpty()) {
```

Comment:
Unknown source with unknown information always looks danger :) But we don't need to worry too much about it because underlying command argument parser will also check. And the launched APP will have same uid with original APP. Even if someone could hack freerdp by some argument (in case we really have bug !-_-), he can't get more permission (I agree there might be some unexpected behavior being allowed unexpectly, though.

On the other hand, bash never do more check for us when someone type 'xfreerdp' in bash console ;)